Config path quote handling: keys with quotes

scheibelp opened the pull request:
Fixes https://github.com/spack/spack/issues/40944

As observed in #40944, when using `spack config add <path>`, the `path` might contain keys that are enclosed in quotes.

This was broken in https://github.com/spack/spack/pull/39831, which assumed that only the value (if present, the final element of the path) would use quotes.

This preserves the primary intended behavior of #39931 (allowing ":" in values when using `spack config add`) while also allowing quotes on keys.

This has complicated the function `process_config_path`, but

* It is not used outside of `config.py`
* The docstring has been updated to account for this

scheibelp committed:
update test according to #40944

scheibelp committed:
fix test; keys are not processed as yaml, so strip quotes

scheibelp committed:
update comment

scheibelp committed:
update comment and style fix

scheibelp committed:
must remove quotes before converting to syaml_str

scheibelp committed:
need to change return type of process_config_path, and add explanation for that

scheibelp committed:
remove typehint; clean comment

scheibelp committed:
rewrite docstring

scheibelp committed:
add edge case test

scheibelp committed:
more docstring editing

haampie commented:
Can the syntax of this amazing ad-hoc language be formalized at least a bit instead of defining it in imperative code?

Something along the lines of:

```python
quoted_string = r'\"[^\"]+?\"'
something_else = r'[^:\"]+?'  # or can you have a " half-way a key that is not a string?
key = rf'{quoted_string}|{something_else}'
list = rf'^({key})(:({key}))*$'  # what about blanks?
```

scheibelp committed:
strict enforcement of dsl

scheibelp committed:
correction in pattern

scheibelp committed:
auto style fix

scheibelp committed:
make use of validator to simplify extraction of tokens

scheibelp committed:
make use of validator to simplify extraction of tokens

scheibelp committed:
improper association of overrides

scheibelp committed:
Revert "improper association of overrides"

This reverts commit b878491c7e689185a66cd869925b26894139987b.

scheibelp committed:
Revert "make use of validator to simplify extraction of tokens"

This reverts commit 9197db00a82157c1b9a5f613098992de3304fd2c.

scheibelp committed:
Revert "make use of validator to simplify extraction of tokens"

This reverts commit e02e5f0fdc6a1da173067e97841ab179c0aeb155.

scheibelp committed:
correction in validation regex

scheibelp committed:
attempt to reform process along regex

scheibelp committed:
fix a couple bugs

scheibelp committed:
seen override

scheibelp committed:
apparently a single : at the end means override

scheibelp committed:
auto style fix

scheibelp commented:
@haampie I added regexes and created an object to formalize the DSL, added a test for that, and refactored parsing to make use of regular expressions as well. Does this look reasonable?

scheibelp committed:
merge from develop (large conflict diff, but seems semantically trivial)

scheibelp committed:
mypy complaints

scheibelp committed:
add (now failing) test

scheibelp committed:
todo comments

scheibelp committed:
changes to validation regex to allow certain values

scheibelp committed:
auto style fix

scheibelp committed:
partial reorg for validation that doesnt lean on a single, large regex

scheibelp committed:
more factoring

scheibelp committed:
more error parsing

scheibelp committed:
single-re validation pattern works

scheibelp committed:
remove stale comment, add test cases

scheibelp committed:
update stale comment (handled in .process vs. .next_token)

scheibelp committed:
alternate strategy

scheibelp committed:
intermediate work on reorg

scheibelp committed:
more intermediate work

scheibelp committed:
better but not 100 percent

scheibelp committed:
better but not 100 percent

scheibelp committed:
now semantics seem reasonable

scheibelp committed:
style fixes

scheibelp committed:
fix last-value processing bug; expand test to check things more thoroughly

scheibelp commented:
This is now resynched (had a merge conflict for a while).

Also, new tests caught an issue with parsing YAML values, so I had to update the parsing and also updated the `config_path_dsl` test with an explicit check. At a higher level, I split the parsing to check if something is either a key or not:

* in the first case, it is covered by a regex
* in the second, it may be a YAML value, but in that case it would have to be the last entry of `x:y:z`, so in that case I attempt to use the YAML handling logic to parse it as such
